libass/ass_font.c

```diff
@@ -35,6 +42,197 @@
 #include "ass_utils.h"
 #include "ass_shaper.h"
 
+
+/**
+ * Convert a UCS-4 code unit to a packed uint32_t in given multibyte encoding
+ */
+#if defined(_WIN32)
+static uint32_t convert_unicode_to_mb(FT_Encoding encoding, uint32_t codepoint)
+{
+    uint32_t result = codepoint;
+    uint16_t input_buffer[2];
```

Comment:
I believe `wchar_t` is the canonical type for `input_buffer` for use with Win32-APIs.

Reply:
FWIW the _actual_ canonical type for Win32 is `WCHAR`, but we use `wchar_t` in libass. This distinction doesn’t matter on native Win32, but on midipix, these types are different. Of course, midipix still has the chance to never take off.

Reply:
Prior to the patches at hand `wchar_t` is only used in `ass_directwrite.c`, which should probably all use `WCHAR` instead.  `dwrite_c.h` already uses `WCHAR` and #573 would currently add more uses of `wchar_t`. Perhaps we should use `WCHAR` where applicable in all new code and also update `ass_directwrite.c`?

Reply:
Yeah, because we don’t currently use Win32 APIs anywhere but in `ass_directwrite.c`. `dwrite_c.h` copies Microsoft’s declarations fairly verbatim, so that explains that.

> Perhaps we should use `WCHAR` where applicable in all new code and also update `ass_directwrite.c`?

Yes, that’s an option. Indeed, that’s the “right” thing to do, probably.

Reply:
> these types are different

If `WCHAR` and `wchar_t` are different, the whole module is likely to break. Since the library functions like `wcsncmp` expect `wchar_t` and Windows APIs expect `WCHAR`. I think we need to place a warning sign here to reminder anyone who want to compile this project on midipix. They will need to use utf-16 variant of the library functions provided by midipix. Or we will need some ifdef-endif to select functions.

Reply:
You mean `ass_directwrite`, right? You’re not using any C library `wchar_t` functions in this PR, are you?

But yes. Good catch. `ass_directwrite` does use `wcsncmp` once. But it can be replaced by a call to `CompareStringW` or `CompareStringEx`, and I don’t immediately see any other `wcs...` calls. But to be safe, a more careful audit of the code would be in order if we were to actually do this.

Reply:
However, `CompareStringOrdinal` is available since Vista. Maybe providing our own version of `wcsncmp` would be better, if we don't want to drop Windows XP.

Reply:
Right, `CompareStringOrdinal` is even better; I overlooked it.

On the one hand, DirectWrite itself is only available on Vista+, so we can safely call that function. On the other hand, we can’t reference the symbol directly lest we fail to launch on XP; we’d have to go via `GetProcAddress` like we do for `DWriteCreateFactory`. That’s kind of annoying for such a simple function used only once.

Another alternative is [`StrCmpW`](https://docs.microsoft.com/en-us/windows/win32/api/shlwapi/nf-shlwapi-strcmpw) from `Shlwapi.dll`, but that means linking to a whole new DLL.

I thought there was a Win32 function for this that had always existed… oh well.

Implementing our own `wcsncmp` is a bulletproof solution, but also feels kinda dumb.

---

Actually, there’s another issue here besides `wcsncmp`: wide string literals. They use `wchar_t`, of course, but we need specifically UTF-16.

At the end of the day, midipix still isn’t exactly a common reality, so we _might_ want to settle for just using `wcsncmp` and wide string literals with `WCHAR` until it materializes better. Or not touch anything at all just yet and stick to `wchar_t` for the time being.

Reply:
So let's just use `wcsncmp` for string comparison and `WCHAR` for where `wchar_t` is needed. This is done by latest commit.

Reply:
Regarding string literals, ISO-C11 has the `u` and `U` prefixes for `char16_t` and `char32_t` respectively. MinGW does not complain about them even with our current setting of `-std=gnu99`.
As of current master the `wcsncmp`-call in `ass_directwrite` appears to be the only standard `wchar_t`-API used. *(Checked by grepping for all functions and macros declared in ISO-C99's `wchar.h`: `-E '(fgetwc|fgetws|fputwc|fputws|fwide|fwprintf|fwscanf|getwc|getwchar|putwc|putwchar|swprintf|swscanf|ungetwc|vfwprintf|vwprintf|wprintf|wscanf|wcstod|wcstof|wcstol|wcstold|wcstoll|wcstoull|wcstoul|btowc|mbrlen|mbrtowc|mbsinit|mbsrtowcs|wcrtomb|wctomb|wcsrtombs|wcscat|wcschr|wcscmp|wcscoll|wcscpy|wcscspn|wcslen|wcncat|wcsncmp|wcsncpy|wcspbrk|wcsrchr|wcsspn|wcsstr|wcstok|wcsxfrm|wmemchr|wmemcmp|wmemcpy|wmemmove|wmemset|wcsftime)\(|WCHAR_'`)*

The non-Unicode-font patches use `wchar_t`/`WCHAR` only with Win32-APIs and no "wide" string- or character-literals or standard "wide" functions; so at least here we can safely use `WCHAR` even if we delay adjusting the other occurrences of `wchar_t` with Win32-APIs.